### PR TITLE
Implement postgresql search

### PR DESCRIPTION
**Pull request by tung2744**
ref DEV-303
ref DEV-1156

How to test:

1. Update your .env according to .env.example
2. Update your authgear.secrets.yaml. Add the search db credential:
  ```
  - data:
      database_schema: public
      database_url: postgres://postgres:postgres@127.0.0.1:5432/search?sslmode=disable
    key: search.db
  ```
3. Run migration `go run ./cmd/authgear search database migrate up`
4. Reindex existing users `go run ./cmd/authgear search reindex {app-id}`

### CODE REVIEW

**Comment by tung2744** [comment]
Added a commit to update the init command to ask for search database credentials

File: cmd/authgear/main.go [hunk 1]
```diff
@@ -19,6 +19,7 @@ import (
 	_ "github.com/authgear/authgear-server/cmd/authgear/cmd/cmdimport"
 	_ "github.com/authgear/authgear-server/cmd/authgear/cmd/cmdinit"
 	_ "github.com/authgear/authgear-server/cmd/authgear/cmd/cmdinternal"
+	_ "github.com/authgear/authgear-server/cmd/authgear/cmd/cmdsearch"
```
Comment (louischan-oursky) [inline cmd/authgear/main.go:22]: Need to do this change in custombuild/authgearx/main.go as well.

File: pkg/lib/search/service.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	case config.SearchImplementationElasticsearch:
+		result, stats, err := s.ElasticsearchService.QueryUser(searchKeyword, filterOptions, sortOption, pageArgs)
+		if err != nil {
+			return nil, nil, err
+		}
+		return result, StatsFromElasticsearch(stats), err
+	case config.SearchImplementationPostgresql:
+		return nil, nil, fmt.Errorf("not implemented")
+	}
+	return nil, nil, fmt.Errorf("unknown search implementation: %s", s.SearchConfig.GetImplementation())
```
Comment (louischan-oursky) [inline pkg/lib/search/service.go:41]: Use panic for unreachable case.

File: pkg/lib/infra/db/searchdb/deps.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+func NewSQLExecutor(handle *Handle) *SQLExecutor {
+	if handle == nil {
+		return nil
+	}
+
+	return &SQLExecutor{
+		db.SQLExecutor{
+			Database: handle,
```
Comment (louischan-oursky) [inline pkg/lib/infra/db/searchdb/deps.go:71]: SQLExecutor no longer has this field `Database`

File: pkg/lib/search/pgsearch/service.go [hunk 4]
```diff
@@ -1,20 +1,110 @@
...
+	ctx context.Context,
 	searchKeyword string,
 	sortOption user.SortOption,
 	pageArgs graphqlutil.PageArgs) ([]apimodel.PageItemRef, error) {
-	return nil, nil
+	if s.SQLExecutor == nil {
+		return nil, fmt.Errorf("search database credential is not provided")
+	}
+	var refs []apimodel.PageItemRef
+	err := s.SQLExecutor.Database.ReadOnly(ctx, func(ctx context.Context) error {
```
Comment (louischan-oursky) [inline pkg/lib/search/pgsearch/service.go]: Inject the handle directly. Do not use the field of a dependency.

File: cmd/authgear/cmd/cmdsearch/reindex.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+				dbPool,
+				config.NewDefaultDatabaseEnvironmentConfig(),
+				searchDatabaseCredentials,
+				loggerFactory)
+			searchdbSQLBuilder := searchdb.NewSQLBuilder(searchDatabaseCredentials)
+			searchdbSQLExecutor := searchdb.NewSQLExecutor(searchdbHandle)
+			store := pgsearch.NewStore(config.AppID(appID), searchdbSQLBuilder, searchdbSQLExecutor)
+
+			reindexer := cmdpgsearch.NewReindexer(dbPool, dbCredentials, cmdpgsearch.CmdAppID(appID))
+			err = searchdbHandle.WithTx(ctx, func(ctx context.Context) error {
```
Comment (louischan-oursky) [inline cmd/authgear/cmd/cmdsearch/reindex.go:77]: If the app contains many users, then this transaction could be opened for a long time. Maybe it is better to let the reindexer to manage the transaction. For now, we can just do 1 txn for 1 source (or user).

File: cmd/authgear/pgsearch/reindexer.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+func (q *Reindexer) reindex(ctx context.Context, store *pgsearch.Store) (allUserIDs map[string]struct{}, err error) {
+	allUserIDs = make(map[string]struct{})
+
+	var first uint64 = 500
+	var after model.PageCursor = ""
+	var items []reindex.ReindexItem
+
+	for {
+		err = q.Handle.WithTx(ctx, func(ctx context.Context) (err error) {
```
Comment (louischan-oursky) [inline cmd/authgear/pgsearch/reindexer.go:51]: I suggest we do not open tx at the same time. Open a tx in appdb to fetch the source, close this tx. And then open a tx in searchdb to upsert.

File: go.mod [hunk 7]
```diff
@@ -93,6 +93,7 @@ require (
 	github.com/authgear/oauthrelyingparty v1.5.0
 	github.com/coder/websocket v1.8.12
 	github.com/go-ldap/ldap/v3 v3.4.8
+	github.com/rivo/uniseg v0.4.7
```
Comment (louischan-oursky) [inline go.mod:96]: Need to update e2e/go.mod e2e/go.sum custombuild/go.mod custombuild/go.sum as well

File: cmd/authgear/config/options.go [hunk 8]
```diff
@@ -83,7 +83,23 @@ func ReadSkipEmailVerification() bool {
 	}.Prompt()
 }
 
-func ReadSecretConfigOptionsFromConsole() *config.GenerateSecretConfigOptions {
+func ReadSearchImplementation() config.SearchImplementation {
+	return config.SearchImplementation(promptString{
```
Comment (louischan-oursky) [inline cmd/authgear/config/options.go:87]: There should be 3 options, `none`, `elasticsearch`, and `postgresql`.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

File: pkg/lib/search/pgsearch/service.go (hunk 4 above)
Comment (tung2744) [reply pkg/lib/search/pgsearch/service.go]: Fixed in later commits

File: go.mod (hunk 7 above)
Comment (tung2744) [reply go.mod:96]: Fixed in later commit

File: cmd/authgear/config/options.go (hunk 8 above)
Comment (louischan-oursky) [reply cmd/authgear/config/options.go:87]:
```
var ErrSearchDisabled = apierrors.InternalError.WithReason("SearchDisabled").New("Search disabled")
```

**Review by louischan-oursky — APPROVED** [review]

**Pull request merged**